## Review fixes pushed

I've pushed fixes for the three open review threads to a sibling branch: [`claude/fix-pr-412-ci-hCs4J`](https://github.com/AgentWorkforce/burn/tree/claude/fix-pr-412-ci-hCs4J) (commit `86d8af9`). The PR author can pull/cherry-pick that commit onto `feat/stamps-export`.

**What changed in `crates/relayburn-cli/src/commands/stamps.rs`:**
- Streams JSONL rows directly into `BufWriter<{StdoutLock, File}>` via `serde_json::to_writer`. No more `Vec<Value>` + concatenated `String` materialization.
- Propagates serialization and I/O errors through `anyhow::Error` → `report_error`, removing the silent `if let Ok(line) = ...` drop. Bad records now fail the export loudly with the right exit code.
- Reports the count actually written, not the input length.

**What changed in `crates/relayburn-cli/tests/smoke.rs`:**
- `stamps_export_to_file_against_empty_ledger` now asserts the output file exists, is empty, and stdout is empty — pins both halves of the `--out <path>` contract.

**Local verification:**
- `cargo test -p relayburn-cli --test smoke -- stamps` — 4/4 pass
- `cargo test --workspace` — full suite green
- `cargo clippy -p relayburn-cli --all-targets` — no new warnings on `stamps.rs`

### About the failing CI

The `build-and-test` (npm tests) job failure is **not** caused by this PR's code. The job fails at install time with:

```
WARN  GET https://registry.npmjs.org/@relayburn/sdk-linux-x64-gnu/-/sdk-linux-x64-gnu-2.7.3.tgz error (ERR_PNPM_TARBALL_INTEGRITY)
```

The just-published `@relayburn/sdk-linux-x64-gnu@2.7.3` tarball on the npm registry has an integrity mismatch, so `pnpm install` can't fetch the optional native dep, and `node --test` then fails to load the napi binding in `@relayburn/mcp` with `native binding not found for linux-x64`. This affects every PR opened after the 2.7.3 publish, not just this one — it's an upstream/release-side issue and would need either (a) a republish of the 2.7.3 native packages or (b) a CI workaround that runs `pnpm run build:napi` before `pnpm run test` so tests don't depend on the published optionalDeps. Out of scope for this PR.

---
_Generated by [Claude Code](https://claude.ai/code)_

--- a/crates/relayburn-cli/src/commands/stamps.rs
+++ b/crates/relayburn-cli/src/commands/stamps.rs
@@ -8,6 +8,9 @@
 //! {"v":1,"kind":"stamp","ts":"2025-01-01T00:00:00Z","selector":{"sessionId":"..."},"enrichment":{"role":"..."}}
 //! ```
 
+use std::fs::File;
+use std::io::{self, BufWriter, Write};
+
 use relayburn_sdk::{ExportStampsOptions, Ledger, LedgerOpenOptions};
 
 use crate::cli::{GlobalArgs, StampsArgs};
@@ -26,7 +29,6 @@ pub fn run(globals: &GlobalArgs, args: StampsArgs) -> i32 {
 fn run_export(globals: &GlobalArgs, args: crate::cli::StampsExportArgs) -> i32 {
     let progress = TaskProgress::new(globals, "stamps export");
 
-    // Open the ledger
     let opts = LedgerOpenOptions {
         home: globals.ledger_path.clone(),
         content_home: None,
@@ -40,46 +42,57 @@ fn run_export(globals: &GlobalArgs, args: crate::cli::StampsExportArgs) -> i32 {
         }
     };
 
-    // Export stamps
     progress.set_task("exporting stamps");
-    let values: Vec<serde_json::Value> = match handle
-        .export_stamps(ExportStampsOptions::default())
-    {
-        Ok(iter) => iter.collect(),
+    let iter = match handle.export_stamps(ExportStampsOptions::default()) {
+        Ok(iter) => iter,
         Err(err) => {
             progress.finish_and_clear();
             return report_error(&err, globals);
         }
     };
+
+    let out_path = args.out.as_deref().unwrap_or(DEFAULT_OUT);
+    let result = if out_path == "-" {
+        let stdout = io::stdout();
+        write_jsonl(&mut BufWriter::new(stdout.lock()), iter)
+    } else {
+        match File::create(out_path) {
+            Ok(file) => write_jsonl(&mut BufWriter::new(file), iter),
+            Err(err) => Err(anyhow::anyhow!("failed to open output file: {}", err)),
+        }
+    };
     progress.finish_and_clear();
 
-    // Determine output destination
-    let out_path = args.out.as_deref().unwrap_or(DEFAULT_OUT);
-
-    // Serialize output
-    let stamp_count = values.len();
-    let mut output = String::new();
-    for val in &values {
-        if let Ok(line) = serde_json::to_string(val) {
-            output.push_str(&line);
-            output.push('\n');
-        }
-    }
-
-    // Write to stdout or file
-    if out_path == "-" {
-        print!("{}", output);
-        0
-    } else {
-        match std::fs::write(out_path, output) {
-            Ok(()) => {
+    match result {
+        Ok(stamp_count) => {
+            if out_path != "-" {
                 eprintln!("Exported {} stamp(s) to {}", stamp_count, out_path);
-                0
             }
-            Err(err) => {
-                let err = anyhow::anyhow!("failed to write output file: {}", err);
-                report_error(&err, globals)
-            }
+            0
         }
+        Err(err) => report_error(&err, globals),
     }
+}
+
+/// Serialize each stamp directly to `writer` as a JSONL line. Returns the
+/// number of records written. Any serialization or I/O error aborts the
+/// export and is propagated — partial files are surfaced as failures so a
+/// `stamps export` user never silently loses records.
+fn write_jsonl<W: Write, I: IntoIterator<Item = serde_json::Value>>(
+    writer: &mut W,
+    iter: I,
+) -> anyhow::Result<usize> {
+    let mut count: usize = 0;
+    for val in iter {
+        serde_json::to_writer(&mut *writer, &val)
+            .map_err(|err| anyhow::anyhow!("failed to serialize stamp: {}", err))?;
+        writer
+            .write_all(b"\n")
+            .map_err(|err| anyhow::anyhow!("failed to write stamp: {}", err))?;
+        count += 1;
+    }
+    writer
+        .flush()
+        .map_err(|err| anyhow::anyhow!("failed to flush output: {}", err))?;
+    Ok(count)
 }

--- a/crates/relayburn-cli/tests/smoke.rs
+++ b/crates/relayburn-cli/tests/smoke.rs
@@ -431,7 +431,20 @@ fn stamps_export_to_file_against_empty_ledger() {
         .env("NO_COLOR", "1")
         .assert()
         .success()
+        .stdout("")
         .stderr(predicate::str::contains("Exported 0 stamp(s)"));
+
+    assert!(
+        out_path.is_file(),
+        "expected output file to be created at {}",
+        out_path.display()
+    );
+    let content = std::fs::read_to_string(&out_path).expect("read exported file");
+    assert!(
+        content.is_empty(),
+        "expected empty JSONL for empty ledger, got {:?}",
+        content
+    );
 }
 
 /// `burn stamps export --json` against an empty ledger should still only

